src/AppInstallerCommonCore/WindowsFeature.cpp

```diff
@@ -0,0 +1,207 @@
+            hr = m_dismEnableFeature(m_session, Utility::ConvertToUTF16(name).c_str(), NULL, DismPackageNone, FALSE, NULL, NULL, TRUE, NULL, NULL, NULL);
+            LOG_IF_FAILED(hr);
+        }
+
+        return hr;
+    }
+
+    HRESULT DismApiHelper::DisableFeature(const std::string_view& name)
+    {
+        HRESULT hr = ERROR_PROC_NOT_FOUND;
+        if (m_dismDisableFeature)
+        {
+            hr = m_dismDisableFeature(m_session, Utility::ConvertToUTF16(name).c_str(), NULL, FALSE, NULL, NULL, NULL);
+            LOG_IF_FAILED(hr);
+        }
+
+        return hr;
+    }
+
+    void DismApiHelper::Initialize()
+    {
+        if (m_dismInitialize)
+        {
+            LOG_IF_FAILED(m_dismInitialize(2, NULL, NULL));
+        }
+    }
+
+    void DismApiHelper::OpenSession()
+    {
+        if (m_dismOpenSession)
+        {
+            LOG_IF_FAILED(m_dismOpenSession(DISM_ONLINE_IMAGE, NULL, NULL, &m_session));
+        }
+    }
+
+    void DismApiHelper::Delete()
+    {
+        if (m_dismDelete)
+        {
+            LOG_IF_FAILED(m_dismDelete(m_featureInfo));
+        }
+    }
+
+    void DismApiHelper::Shutdown()
+    {
+        if (m_dismShutdown)
+        {
+            LOG_IF_FAILED(m_dismShutdown());
+        }
+    }
+
+#ifndef AICLI_DISABLE_TEST_HOOKS
+    static HRESULT* s_EnableWindowsFeatureResult_TestHook_Override = nullptr;
+
+    void TestHook_SetEnableWindowsFeatureResult_Override(HRESULT* result)
+    {
+        s_EnableWindowsFeatureResult_TestHook_Override = result;
+    }
+#endif
+
+    HRESULT WindowsFeature::EnableFeature()
+    {
+#ifndef AICLI_DISABLE_TEST_HOOKS
+        if (s_EnableWindowsFeatureResult_TestHook_Override)
+        {
+            return *s_EnableWindowsFeatureResult_TestHook_Override;
+        }
+#endif
+        return m_dismApiHelper.EnableFeature(m_featureName);
+    }
+
+    HRESULT WindowsFeature::DisableFeature()
+    {
+        return m_dismApiHelper.DisableFeature(m_featureName);
+    }
+
+    bool WindowsFeature::DoesExist()
+    {
+        return m_dismApiHelper.GetFeatureInfo(m_featureName);
+    }
+
+#ifndef AICLI_DISABLE_TEST_HOOKS
+    static bool* s_IsWindowsFeatureEnabledResult_TestHook_Override = nullptr;
+
+    void TestHook_SetIsWindowsFeatureEnabledResult_Override(bool* status)
+    {
+        s_IsWindowsFeatureEnabledResult_TestHook_Override = status;
+    }
+#endif
+
+    bool WindowsFeature::IsEnabled()
+    {
+#ifndef AICLI_DISABLE_TEST_HOOKS
+        if (s_IsWindowsFeatureEnabledResult_TestHook_Override)
+        {
+            return *s_IsWindowsFeatureEnabledResult_TestHook_Override;
+        }
+#endif
+        DismFeatureInfo* featureInfo = m_dismApiHelper.GetFeatureInfo(m_featureName);
+        DismPackageFeatureState featureState = featureInfo->FeatureState;
+        AICLI_LOG(Core, Verbose, << "Feature state of " << m_featureName << " is " << featureState);
+        return (featureState == DismStateInstalled || featureState == DismStateInstallPending);

```

Comment:
Do you have more information on why `DismStateInstallPending` should be included but `DismStatePartiallyInstalled` is not? My gut would be that only `DismStateInstalled` would be installed and everything else should require us to do something.

Reply:
I changed the condition so that only `DismStateInstalled` is considered as enabled.